### PR TITLE
#341 Introduce TestCafe

### PR DESCRIPTION
**Pull request by josephj**
#### Description

Ability to use TestCafe as our E2E testing tool

#### QA Steps

* Checkout to this `#341` branch on your local machine.
* Install the new required NPM packages with `yarn install`.
* Execute `yarn test:functional` to see if it can execute test **on QA env** successfully.
* Execute `NODE_ENV=development yarn test:functional` to see if it can execute test **on local env** 
successfully. (Need to `yarn start` to have the local server first)
(use `set NODE_ENV=development&&yarn test:functional` on windows)

#### Acceptance Criteria

* Ensure the testing process is working smoothly as you expected
* Ensure the test case itself is easy to understand.

### CODE REVIEW

File: test/e2e-test/index.fixture.js [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    .selectText(NoteInput)
+    .typeText(NoteInput, 'Footnote Test')
+    .click(SaveButton)
+    .expect(FirstCell.innerText)
+    .contains('Footnote Test', 'Modify the first cell to "Footnote Test"');
+
+  await t
+    .click(FirstCell)
+    .click(SwitchButton)
+    .selectText(ReasonInput)
```
Comment (trust2065) [inline test/e2e-test/index.fixture.js]: The test will fail if the SwitchButton was 'on' and the test make it 'off'. Hence the test can't find the input.
image

**Comment by josephj** [comment]
@trust2065 

I feel it's probably because that TestChrome can't find the Headless Chrome on your system. I will try to specify one with [puppeteer](https://github.com/GoogleChrome/puppeteer).

File: test/e2e-test/index.fixture.js (hunk 1 above)
Comment (josephj) [reply test/e2e-test/index.fixture.js]: Good catch. I have reset the data if it's necessary.

**Comment by josephj** [comment]
@trust2065 

I've done the necessary changes. Please help to check it again.

**Comment by trust2065** [comment]
Tried it with puppeteer but the test give the same feedback - hanging without any movement.

**Comment by josephj** [comment]
@trust2065 

Can you try this?

```
npm install -g testcafe;
testcafe "chrome:C:\Users\YourUserName\AppData\Local\Google\Chrome SxS\Application\chrome.exe:headless" test.js` 
```

**Comment by trust2065** [comment]
@josephj 

Somehow I can't run the test with the full path of browser.
But I am able to run the code within 4s with
`testcafe "chrome" test/e2e-test/test.js`

For headless, it's still not working
`testcafe "chrome:headless" test/e2e-test/test.js`

**Pull request merged**